Save associated_files to xml

mesmith75 opened the pull request:
This writes the list of `associated_files` from a `ShareDir` to an XML file in the shared directory. This means that if a job is copied and new files added, the original will know about them.

Also added to `GaudiExec` to check the existence of previously uploaded files in case users have been cleaning their grid areas between job submissions.

rob-c commented on python/Ganga/GPIDev/Lib/File/File.py:
```diff
@@ -296,13 +300,30 @@ def isExecutable(self):
 
+    def getAssociatedFiles(self):
+        if os.path.isfile(os.path.join(self.path(), 'associated_files.xml')):
+            fobj = open(os.path.join(self.path(), 'associated_files.xml'), "r")
```
I forget the naming but isn't this safer as `with open( ... ) :` rather than opening and closing the file handler?

mesmith75 replied:
good point.

mesmith75 commented:
@egede yes - the `safe_save` has a lock in it but I will add to the other functions. Is the `synchronised` decorator sufficient here?

mesmith75 commented:
I have added in the `@synchronised` decorator which I think should provide the necessary protection.

I have also added to the ShareRef to remove the `sharedir` from the list when the counter gets to 0. Currently old deleted sharedirs hang about in the prep registry.

rob-c commented:
@mesmith75 I'm only asking as I can't remember how this works now (been a while since I looked in detail at this).
Isn't the list of LFN which have been uploaded for a `SharedDir` saved with the same object on disk now? If that's the case isn't the object copied when the `ShareDir` object is copied?

I'm just left wondering why this is implemented this way but I could well be missing something.

mesmith75 commented:
@rob-c Kind of - but the ShareDir is saved with job. If you copy a job then the new job has a copy of the `ShareDir` object so if you add an LFN to the copy the original doesn't know about it.

rob-c commented:
@mesmith75 I wonder if there is a mixing of temporary prepared LFN per job and shared LFN which don't change between jobs.
Would a non-copyable list attribute of ShareDir be better as the original job shouldn't know about future jobs files and if there is some prepared data specific to an individual job it shouldn't be exposed to future jobs I would have suspected?

mesmith75 commented:
@rob-c I see your point - I guess technically this is not correct for the `ShareDir`. 

The thinking is that if you copy a GaudiExec job you end up with a new LFN for the job scripts which needs to be removed - but maybe that should be removed with the job rather than the `ShareDir` as it is specific to the job in question.

But also you may upload a new cmake tarball in case the old one has disappeared (which is fine as it is just uploading what was in the local `ShareDir` so the application itself will still be the same). For that it is a bit trickier because the new upload can still be copied for new jobs so it does need to be part of the `ShareDir` object.
